### PR TITLE
Set only compressible resources on system and kube reserved cgroup slices

### PR DESCRIPTION
**Pull request by harche**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

A few weeks ago in the sig-node meeting we talked about tackling this long standing issue, https://github.com/kubernetes/kubernetes/issues/72881. This issue causes system reserved to get allocated incorrect amount of resources than indicated by the user when specifying system reserved cpu values. Detailed analysis with various CPUs is documented here, https://docs.google.com/spreadsheets/d/1N8Xkzu7ArZYKTP0Ob9vMxUnlqAFlsaFoNGbxyAhzzxU/edit?gid=1029636607#gid=1029636607 

While the enforceNodeAllocatable solves the issue for the CPU, it was brought to our attention by @derekwaynecarr that setting it for memory has unintended consequences. Hence, it was proposed that we come up with a way of setting only compressible (e.g. CPU) resources on the system slice to mitigate the issue. 

This PR introduces `system-reserved-compressible` and `kube-reserved-compressible` to set only compressible (e.g. CPU) type resources on system and kube reserved slice respectively.  

Here is a simple example, 
With the `500m` cpu and `1Gi` memory for system reserved the respective cgroup slices look like this,

```bash
# cat system.slice/cpu.weight
100
# cat kubepods.slice/cpu.weight
137
```
Now with the changes in this PR if we set the `system-reserved-compressible` using,
```yaml
kubeletConfig:
    systemReservedCgroup: "/system.slice"
    enforceNodeAllocatable:
      - pods
      - system-reserved-compressible
    systemReserved:
      cpu: "500m"
      memory: "1Gi"
```
those slices now look like,

```bash
# cat system.slice/cpu.weight
20
# cat kubepods.slice/cpu.weight
137
```

#### Which issue(s) this PR fixes:

Fixes # https://github.com/kubernetes/kubernetes/issues/72881

#### Does this PR introduce a user-facing change?

```release-note
Introduce compressible resource setting on system reserved and kube reserved slices
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
Introduce compressible resource setting on system reserved and kube reserved slices

```

### CODE REVIEW

File: pkg/kubelet/apis/config/validation/validation.go [hunk 1]
```diff
@@ -203,6 +203,41 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
...
+			reservedKeys[val] = true
+
+		case kubetypes.KubeReservedCompressibleEnforcementKey:
+			if reservedKeys[kubetypes.KubeReservedEnforcementKey] {
+				allErrors = append(allErrors, fmt.Errorf("invalid configuration: both %q and %q cannot be specified together in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.KubeReservedEnforcementKey, kubetypes.KubeReservedCompressibleEnforcementKey))
+			}
+			reservedKeys[val] = true
+		}
+	}
+
```
Comment (mrunalp) [inline pkg/kubelet/apis/config/validation/validation.go:240]:
```suggestion
	for _, val := range kc.EnforceNodeAllocatable {
		reservedKeys[val] = true
	}

	if reservedKeys[kubetypes.SystemReservedCompressibleEnforcementKey] && reservedKeys[kubetypes.SystemReservedEnforcementKey] {
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: both %q and %q cannot be specified together in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey))
	}

	if reservedKeys[kubetypes.KubeReservedCompressibleEnforcementKey] && reservedKeys[kubetypes.KubeReservedEnforcementKey] {
		allErrors = append(allErrors, fmt.Errorf("invalid configuration: both %q and %q cannot be specified together in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.KubeReservedEnforcementKey, kubetypes.KubeReservedCompressibleEnforcementKey))
	}
```

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -498,7 +498,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
...
-	fs.StringSliceVar(&c.EnforceNodeAllocatable, "enforce-node-allocatable", c.EnforceNodeAllocatable, "A comma separated list of levels of node allocatable enforcement to be enforced by kubelet. Acceptable options are 'none', 'pods', 'system-reserved', and 'kube-reserved'. If the latter two options are specified, '--system-reserved-cgroup' and '--kube-reserved-cgroup' must also be set, respectively. If 'none' is specified, no additional options should be set. See https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/ for more details.")
+	fs.StringSliceVar(&c.EnforceNodeAllocatable, "enforce-node-allocatable", c.EnforceNodeAllocatable, "A comma separated list of levels of node allocatable enforcement to be enforced by kubelet. Acceptable options are 'none', 'pods', 'system-reserved', 'system-reserved-compressible', 'kube-reserved' and 'kube-reserved-compressible'. If the latter four options are specified, '--system-reserved-cgroup' and '--kube-reserved-cgroup' must also be set, respectively. If 'none' is specified, no additional options should be set. See https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/ for more details.")
```
Comment (mrunalp) [inline cmd/kubelet/app/options/options.go:501]:
```suggestion
	fs.StringSliceVar(&c.EnforceNodeAllocatable, "enforce-node-allocatable", c.EnforceNodeAllocatable, "A comma separated list of levels of node allocatable enforcement to be enforced by kubelet. Acceptable options are 'none', 'pods', 'system-reserved', 'system-reserved-compressible', 'kube-reserved' and 'kube-reserved-compressible'. If any of the latter four options are specified, '--system-reserved-cgroup' and '--kube-reserved-cgroup' must also be set, respectively. If 'none' is specified, no additional options should be set. See https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/ for more details.")
```

File: pkg/kubelet/apis/config/validation/validation.go [hunk 3]
```diff
@@ -213,13 +233,13 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
 
 		switch val {
 		case kubetypes.NodeAllocatableEnforcementKey:
-		case kubetypes.SystemReservedEnforcementKey:
+		case kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey:
 			if kc.SystemReservedCgroup == "" {
-				allErrors = append(allErrors, fmt.Errorf("invalid configuration: systemReservedCgroup (--system-reserved-cgroup) must be specified when %q contained in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey))
+				allErrors = append(allErrors, fmt.Errorf("invalid configuration: systemReservedCgroup (--system-reserved-cgroup) must be specified when %q or %q contained in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey))
```
Comment (mrunalp) [inline pkg/kubelet/apis/config/validation/validation.go:238]:
```suggestion
				allErrors = append(allErrors, fmt.Errorf("invalid configuration: systemReservedCgroup (--system-reserved-cgroup) must be specified when %q or %q included in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey))
```

File: pkg/kubelet/apis/config/validation/validation.go [hunk 4]
```diff
@@ -213,13 +233,13 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
...
 			}
-		case kubetypes.KubeReservedEnforcementKey:
+		case kubetypes.KubeReservedEnforcementKey, kubetypes.KubeReservedCompressibleEnforcementKey:
 			if kc.KubeReservedCgroup == "" {
-				allErrors = append(allErrors, fmt.Errorf("invalid configuration: kubeReservedCgroup (--kube-reserved-cgroup) must be specified when %q contained in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.KubeReservedEnforcementKey))
+				allErrors = append(allErrors, fmt.Errorf("invalid configuration: kubeReservedCgroup (--kube-reserved-cgroup) must be specified when %q or %q contained in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.KubeReservedEnforcementKey, kubetypes.KubeReservedCompressibleEnforcementKey))
```
Comment (mrunalp) [inline pkg/kubelet/apis/config/validation/validation.go:242]:
```suggestion
				allErrors = append(allErrors, fmt.Errorf("invalid configuration: kubeReservedCgroup (--kube-reserved-cgroup) must be specified when %q or %q included in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.KubeReservedEnforcementKey, kubetypes.KubeReservedCompressibleEnforcementKey))
```

File: pkg/kubelet/cm/node_container_manager_linux.go [hunk 5]
```diff
@@ -162,22 +183,35 @@ func enforceExistingCgroup(cgroupManager CgroupManager, cName CgroupName, rl v1.
 }
 
 // getCgroupConfig returns a ResourceConfig object that can be used to create or update cgroups via CgroupManager interface.
-func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
+func getCgroupConfig(rl v1.ResourceList, compressibleResources bool) *ResourceConfig {
```
Comment (mrunalp) [inline pkg/kubelet/cm/node_container_manager_linux.go:186]:
```suggestion
func getCgroupConfig(rl v1.ResourceList, compressibleResourcesOnly bool) *ResourceConfig {
```

File: pkg/kubelet/cm/node_container_manager_linux.go [hunk 6]
```diff
@@ -162,22 +183,35 @@ func enforceExistingCgroup(cgroupManager CgroupManager, cName CgroupName, rl v1.
...
+
+	setCompressibleResources := func() {
+		if q, exists := rl[v1.ResourceCPU]; exists {
+			// CPU is defined in milli-cores.
+			val := MilliCPUToShares(q.MilliValue())
+			rc.CPUShares = &val
+		}
+	}
+
+	// Only return compressible resources
```
Comment (mrunalp) [inline pkg/kubelet/cm/node_container_manager_linux.go:201]:
```
	if compressibleResourcesOnly {
		setCompressibleResources()
	} else {
		if q, exists := rl[v1.ResourceMemory]; exists {
			// Memory is defined in bytes.
			val := q.Value()
			rc.Memory = &val
		}

		setCompressibleResources()

		if q, exists := rl[pidlimit.PIDs]; exists {
			val := q.Value()
			rc.PidsLimit = &val
		}
		rc.HugePageLimit = HugePageLimits(rl)
	}

	return &rc
```

**Review by mrunalp — APPROVED** [review]
/lgtm

File: pkg/kubelet/apis/config/validation/validation.go [hunk 7]
```diff
@@ -203,6 +203,26 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
...
+		kubetypes.SystemReservedCompressibleEnforcementKey: false,
+		kubetypes.KubeReservedEnforcementKey:               false,
+		kubetypes.KubeReservedCompressibleEnforcementKey:   false,
+	}
+
+	for _, val := range kc.EnforceNodeAllocatable {
+		reservedKeys[val] = true
+	}
+
+	if reservedKeys[kubetypes.SystemReservedCompressibleEnforcementKey] && reservedKeys[kubetypes.SystemReservedEnforcementKey] {
```
Comment (deads2k) [inline pkg/kubelet/apis/config/validation/validation.go:218]: prior to this change, what did the kubelet do when these were both set?

File: pkg/kubelet/apis/config/validation/validation.go [hunk 8]
```diff
@@ -203,6 +203,26 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
...
+		allErrors = append(allErrors, fmt.Errorf("invalid configuration: both %q and %q cannot be specified together in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey))
+	}
+
+	if reservedKeys[kubetypes.KubeReservedCompressibleEnforcementKey] && reservedKeys[kubetypes.KubeReservedEnforcementKey] {
```
Comment (deads2k) [inline pkg/kubelet/apis/config/validation/validation.go:222]: prior to this change, what did the kubelet do when these were both set?

File: pkg/kubelet/apis/config/validation/validation.go [hunk 9]
```diff
@@ -213,13 +233,13 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
 
 		switch val {
 		case kubetypes.NodeAllocatableEnforcementKey:
-		case kubetypes.SystemReservedEnforcementKey:
+		case kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey:
```
Comment (deads2k) [inline pkg/kubelet/apis/config/validation/validation.go:236]: prior to this change, what did the kubelet do?

File: pkg/kubelet/apis/config/validation/validation.go [hunk 10]
```diff
@@ -213,13 +233,13 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration, featur
...
 			if kc.SystemReservedCgroup == "" {
-				allErrors = append(allErrors, fmt.Errorf("invalid configuration: systemReservedCgroup (--system-reserved-cgroup) must be specified when %q contained in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey))
+				allErrors = append(allErrors, fmt.Errorf("invalid configuration: systemReservedCgroup (--system-reserved-cgroup) must be specified when %q or %q included in enforceNodeAllocatable (--enforce-node-allocatable)", kubetypes.SystemReservedEnforcementKey, kubetypes.SystemReservedCompressibleEnforcementKey))
 			}
-		case kubetypes.KubeReservedEnforcementKey:
+		case kubetypes.KubeReservedEnforcementKey, kubetypes.KubeReservedCompressibleEnforcementKey:
```
Comment (deads2k) [inline pkg/kubelet/apis/config/validation/validation.go:240]: prior to this change, what did the kubelet do when `KubeReservedCgroup` was missing for `KubeReservedCompressibleEnforcementKey` ?

**Comment by deads2k** [comment]
validation for new fields looks good

/approve

**Pull request merged**